imagepuller/internal/service/service.go

```diff
@@ -0,0 +1,142 @@
+	}()
+
+	ref, err := name.ParseReference(r.ImageUrl)
+	if err != nil {
+		return nil, fmt.Errorf("parsing the image URL as a references: %w", err)
+	}
+
+	tr := transport.NewRetry(remote.DefaultTransport)
+	remoteImg, err := remote.Image(ref, remote.WithContext(ctx), remote.WithTransport(tr))
+	if errors.Is(err, context.Canceled) || errors.Is(ctx.Err(), context.Canceled) {
+		return nil, fmt.Errorf("pull aborted (context cancelled): %w, %w", ctx.Err(), err)
+	} else if err != nil {
+		return nil, fmt.Errorf("accessing the remote image URL: %w", err)
+	}
+
+	layers, err := remoteImg.Layers()
+	if err != nil {
+		return nil, fmt.Errorf("obtaining the image layers: %w", err)
+	}
+
+	store, err := storage.GetStore(types.StoreOptions{
+		TransientStore: true,
+		RunRoot:        filepath.Join(os.TempDir(), "run"),
+		GraphRoot:      filepath.Join(os.TempDir(), "graph"),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("opening storage: %w", err)
+	}
+
+	var parentIDs []string
+	for idx, layer := range layers {
+		rc, err := layer.Compressed() // Uncompressed?
+		if err != nil {
+			return nil, fmt.Errorf(fmt.Sprintf("reading layer %d", idx), err)
+		}
+
+		// Wrap in func to close each rc immediately
+		err = func() error {
+			defer rc.Close()
+			putLayer, n, err := store.PutLayer(
+				"",        // empty ID -> let store decide
+				"",        // parent == top of parentIDs (empty for first)
+				parentIDs, // full parent chain
+				"",        // mount label
+				false,     // readonly
+				nil,       // mount options
+				rc,        // tar stream
+			)
+			if err != nil {
+				return err
+			}
+			log.Info("Applied layer", "id", putLayer.ID, "bytes", n)
+			parentIDs = append(parentIDs, putLayer.ID)
+			return nil
+		}()
+		if errors.Is(err, context.Canceled) || errors.Is(ctx.Err(), context.Canceled) {
+			return nil, fmt.Errorf("pull aborted while applying layer (context cancelled): %w, context error: %w", err, ctx.Err())
+		} else if err != nil {
+			return nil, fmt.Errorf("applying layer: %w", err)
+		}
+	}
+
+	if len(parentIDs) == 0 {
+		return nil, fmt.Errorf("no layers found (empty image?)")
+	}
+
+	newImg, err := store.CreateImage("", nil, parentIDs[len(parentIDs)-1], "", nil)
+	if err != nil {
+		return nil, fmt.Errorf("creating image: %w", err)
+	}
+	log.Info("Created image", "id", newImg.ID)
+
+	container, err := store.CreateContainer("", nil, newImg.ID, "", "", nil)
+	if err != nil {
+		return nil, fmt.Errorf("creating container: %w", err)
+	}
+	log.Info("Created container", "id", container.ID)
+
+	mountPoint, err := store.Mount(container.ID, "")
+	if err != nil {
+		return nil, fmt.Errorf("mounting container: %w", err)
+	}
+	log.Debug("Mounted in tmpdir", "mountPoint", mountPoint)
+
+	mode := os.FileMode(int(755))
+	imagePath := filepath.Join(r.BundlePath, "rootfs")
+	if err := os.MkdirAll(imagePath, mode); err != nil {
+		return nil, fmt.Errorf("creating bundle path: %w", err)
+	}
+	if err := unix.Unmount(imagePath, 0); err == nil {
```

Comment:
Why is that necessary? Not sure whether we should care about what the agent did so far to the rootfs.

Reply:
Alright. Should we just mount over whatever might be mounted, or return an error?